### ISSUE TITLE
OpenCL spatial convolution bug 

### ISSUE THREAD

**Issue by ngaloppo** [BVLC/caffe#4304]
Hardware: SKL i5-6600U - Intel GPU
OS: Ubuntu 16.04

Testing AlexNet (batch 128), as defined by soumith's convnet benchmarks. Also tried 32

```
ViennaCL: Getting queue for device Intel(R) HD Graphics in context 0xbaa860
ViennaCL: Current queue id 0
ViennaCL: Setting handle kernel argument 0x10f22d0 at pos 0 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting int precision kernel argument 0 at pos 1 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting handle kernel argument 0x104e2d0 at pos 2 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting int precision kernel argument 0 at pos 3 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting handle kernel argument 0x2b48d80 at pos 4 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting int precision kernel argument 0 at pos 5 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting handle kernel argument 0x105ee20 at pos 6 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Setting int precision kernel argument 17775936 at pos 7 for kernel U5_5_64_1_1_1_1_31_31_128_1_192_1_4_3_1_2
ViennaCL: Getting queue for device Intel(R) HD Graphics in context 0xbaa860
ViennaCL: Current queue id 0
ViennaCL: Getting current_context with id 0
ViennaCL: Getting queue for device Intel(R) HD Graphics in context 0xbaa860
ViennaCL: Current queue id 0
ViennaCL: Getting queue for device Intel(R) HD Graphics in context 0xbaa860
ViennaCL: Current queue id 0
ViennaCL: Getting queue for device Intel(R) HD Graphics in context 0xbaa860
ViennaCL: Current queue id 0
*** Aborted at 1465841354 (unix time) try "date -d @1465841354" if you are using GNU date ***
PC: @     0x7f2c735c1c03 caffe::ConvolutionLayerSpatial<>::verify_result()
*** SIGSEGV (@0x275) received by PID 3065 (TID 0x7f2c73ad47c0) from PID 629; stack trace: ***
    @     0x7f2c717f44a0 (unknown)
    @     0x7f2c735c1c03 caffe::ConvolutionLayerSpatial<>::verify_result()
    @     0x7f2c735c757b caffe::ConvolutionLayerSpatial<>::setup_convolution()
    @     0x7f2c735c8170 caffe::ConvolutionLayerSpatial<>::Forward_gpu()
    @     0x7f2c7334ce82 caffe::Net<>::ForwardFromTo()
    @     0x7f2c7334cf87 caffe::Net<>::Forward()
    @           0x41388e time()
    @           0x40ebfb main
    @     0x7f2c717df830 __libc_start_main
    @           0x40f469 _start
    @                0x0 (unknown)
Segmentation fault (core dumped)
```

**Comment by cepiross** [comment]
@naibaf7, Relating to Intel's spatial convolution, how do you think about a functionality of either turning on or off this feature at compile time?

In case of Mac OS X, OpenCL driver does not support Intel's extension regardless of Iris graphics, I considered to check platform vendor at layer_factory.cpp. But, it is not a good idea because vieannaCL API only provides vendorID in digits. To check actual platform vendor information, I have to call get_platforms(), compare platform_index() from ctx against a list of platforms and filter Apple label from info() string. Due to inefficiency, I just comment out [layer_factory's invocation](https://github.com/BVLC/caffe/blob/opencl/src/caffe/layer_factory.cpp#L78-81) and have used it until now.

**Comment by naibaf7** [comment]
@cepiross 
Good point, we'll do that.

@gongzg
I was able to verify the bug here, seems like the new CPU verify_result() can segfault in some instances, for example when running: ./build/test/test_all.testbin --gtest_filter=*TestGradient_Spatial*

**Comment by gongzg** [comment]
@cepiross @naibaf7  I agree. It's hard to determine whether the subgroup extension is supported by only check the vendor id. The best way to do this type of thing is the runtime library should provide correct extension strings, then we can check the extension string easily at runtime. But unfortunately, Intel OpenCL SDK doesn't export any of the subgroup extension string.

@ngaloppo @naibaf7 I can't reproduce the bug at my side currently. But I do found some potential risk cases. I will submit a new PR latter, and please have a try to check whether it works at your side. Thanks.

**Comment by gongzg** [comment]
@naibaf7 @ngaloppo Could you try PR 4311 to verify whether it could fix this issue. It's caused by MKL has some issue in the verification code and lead to all kernel configuration fail the verification which trigger a bug in the verification function.

**Comment by ngaloppo** [comment]
@gongzg @naibaf7 @cepiross How about checking for the `cl_intel_subgroups` vendor extension string? Wouldn't that be the best way to check for support for the subgroup functionality? My driver supports this string.

**Comment by gongzg** [comment]
@ngaloppo At the beginning, I tried to use the extension string but failed. I just checked again with the current OpenCL SDK library and found the latest official OpenCL SDK already give the correct extension of cl_intel_subgroups and only the open source beignet driver doesn't provide correct extension string. That's not a big deal. I will contact beignet developer to fix this problem soon. After that, I will submit a new patch to use the extention to determine whether we use Intel Spatial. @naibaf7 @cepiross What do you think?

**Comment by cepiross** [comment]
@ngaloppo @gongzg  if `cl_intel_subgroups` is presented correctly from extensions(), this is the best option. I definitely agree with you.

As soon as the patch is adapted, it is possible to withdraw `USE_INTEL_SPATIAL` option only if compatible kernel works well, I think. Another option is to make spatial convolution optional. It depends on your decision, @naibaf7.

**Comment by naibaf7** [comment]
@cepiross 
I am investigating methods to provide either a common interface for the convolution backends and/or a better backend selection logic. At the moment, the Intel spatial convolutions would also be selected for CPU devices from Intel using OpenCL. This is not intended either, so due to these behaviours I decided to make all convolution backends other than the Caffe default one optional.

Which is not a good solution since many will have sub-par performance if they don't know which backend (libdnn, conv spatial) they have to enable for their device. It gets even worse when multiple devices are present at the same time.
